src/imitation/data/types.py

```diff
@@ -298,24 +297,105 @@ def __post_init__(self):
         _rews_validation(self.rews, self.acts)
 
 
-def load(path: AnyPath) -> Sequence[TrajectoryWithRew]:
+def load_with_rewards(path: AnyPath) -> Sequence[TrajectoryWithRew]:
+    """Loads a sequence of trajectories with rewards from a file."""
+    data = load(path)
+    if not all(isinstance(traj, TrajectoryWithRew) for traj in data):
+        raise ValueError(
+            f"Expected all trajectories to be of type `TrajectoryWithRew`, "
+            f"but found {type(data).__name__}",
+        )
+
+    return cast(Sequence[TrajectoryWithRew], data)
+
+
+def load(path: AnyPath) -> Sequence[Trajectory]:
     """Loads a sequence of trajectories saved by `save()` from `path`."""
-    with open(path, "rb") as f:
-        return pickle.load(f)
+    # Interestingly, np.load will just silently load a normal pickle file when you
+    # set `allow_pickle=True`. So this call should succeed for both the new compressed
+    # .npz format and the old pickle based format. To tell the difference we need to
+    # look at the type of the resulting object. If it's the new compressed format,
+    # it should be a Mapping that we need to decode, whereas if it's the old format
+    # it's just the sequence of trajectories and we can return it directly.
+    data = np.load(path, allow_pickle=True)
+    if isinstance(data, Sequence):
+        return data
+
+    elif isinstance(data, Mapping):
+        num_trajs = len(data["indices"])
+        fields = (
+            # Account for the extra obs in each trajectory
+            np.split(data["obs"], data["indices"] + np.arange(num_trajs) + 1),
+            np.split(data["acts"], data["indices"]),
+            np.split(data["infos"], data["indices"]),
+            data["terminal"],
+        )
+        if "rews" in data:
+            fields += (np.split(data["rews"], data["indices"]),)
+            return [TrajectoryWithRew(*args) for args in zip(*fields)]
```

Comment:
Hmm I did have that thought, but I decided against it since `obs` needs slightly different indices and `terminal` doesn't need to be split at all. Also transposing a dictionary of lists to a list of dictionaries is a bit hackier/less readable than `zip(*fields)` ([you'd have to do](https://stackoverflow.com/questions/5558418/list-of-dicts-to-from-dict-of-lists) `[dict(zip(DL,t)) for t in zip(*DL.values())]` or something similar).

If we want to go for the Mapping approach anyway, one thing that I think would work is this:
```
elif isinstance(data, Mapping):
        num_trajs = len(data["indices"])
	# Account for the extra obs in each trajectory
        data["obs"] = np.split(data["obs"], data["indices"] + np.arange(num_trajs) + 1)
	data["acts"] = np.split(data["acts"], data["indices"])
	data["infos"] = np.split(data["infos"], data["indices"])
	if "rews" in data:
		data["rews"] = np.split(data["rews"], data["indices"])
		transposed = [dict(zip(data, t)) for t in zip(*data.values())]
		return [TrajectoryWithRew(**kwargs) for kwargs in transposed]
	
	transposed = [dict(zip(data, t)) for t in zip(*data.values())]
	return [Trajectory(**kwargs) for kwargs in transposed]
```

Reply:
I was proposing mapping rather than fields so that we're not dependent on positional ordering of `Trajectory[WithRew]`. But transposing dictionary of lists does seem more involved, I think I'm convinced to keep it as-is.